In [65]:
import os, sys, glob, csv
# os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
# The GPU id to use, usually either "0" or "1"
os.environ["CUDA_VISIBLE_DEVICES"]="1" 
import numpy as np
import pandas as pd
import shutil
import pickle
import tqdm
from tqdm import tqdm
import xlsxwriter
import random
from random import shuffle
from math import log, floor
import re
import collections
from collections import Counter
import string
import unicodedata as udata
# import pause, datetime
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
from distutils.dir_util import copy_tree
import sklearn
from sklearn.metrics import *
import itertools as it
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import wordpunct_tokenize
from nltk import word_tokenize
from nltk.stem.porter import *
from nltk.tokenize import RegexpTokenizer
from collections import Counter
import functools
import spacy

import tensorflow.keras.preprocessing.text as T

stop words

In [11]:
stop_words_1 = (stopwords.words('english')) #Stopword
with open('data/preprocess/stop_words.txt') as f:
    stop_words_2 = f.read().splitlines() #stop_list1
stop_words_3 = pickle.load(open('data/preprocess/stop_list2.pkl','rb'))
stop_words_all = set(stop_words_1 + stop_words_2 + stop_words_3)
len(stop_words_all) # stopwords#

594

In [91]:
kk = ['A','B']
[x.lower() for x in kk]
'A5555'.isdigit()

False

In [137]:
set(['A','a'])

{'A', 'a'}

In [86]:
sp_tokenizer = spacy.load("en")
kk = sp_tokenizer("It's (beautiful) 4toys doesn't")
token_li=[]
for token in kk:
    token_li.append(token.text)
token_li

['It', "'s", '(', 'beautiful', ')', '4toys', 'does', "n't"]

Prepare preprocess

In [141]:
sp_tokenizer = spacy.load("en")
dil= r"[!\"#$%&()*+,-./:;<=>?@[\\]^_`{|}~]+\ *" 
dil_filter = '!\"#$%&()*+,-./:;<=>?@[\\]^_\`{|}~\t\n'
glove = pd.read_table('./data/preprocess/glove.840B.300d.txt', sep=" ", index_col=0, header=None, quoting=csv.QUOTE_NONE)
glove_words = set(glove.index.tolist())
len(glove_words)

2196009

In [142]:
def text_preprcess(text,tokenizers='nltk'):
    '''
    Input: string
    Return: preprocessed list
    '''
    tokens = [i for i in text.split() if i not in stop_words_all]
    tokens_str = ' '.join(tokens)
    if tokenizers == 'keras':
        tokens = T.text_to_word_sequence(tokens_str,filters=dil_filter) # lower casse & tokenize
        tokens = list(filter(None, tokens))
    elif tokenizers == 'nltk':
#         tokens = re.sub(dil," ",tokens_str.lower())
        tokens = [i for i in tokens_str.lower() if i not in dil_filter]
        tokens = ''.join(tokens)
        tokens = word_tokenize(tokens)
        tokens = list(filter(None, tokens))
#         tokens = [x.lower() for x in tokens]
    elif tokenizers == 'spacy':
#         tokens = re.sub(dil," ",tokens_str.lower())
        tokens = [i for i in tokens_str.lower() if i not in dil_filter]
        tokens = ''.join(tokens)
        tokens = sp_tokenizer(tokens)
        token_li = []
        for token in tokens:
            token_li.append(token.text)
        tokens = list(filter(None, token_li))
#         tokens = [x.lower() for x in token_li]
    tokens_li = []
    for token in tokens:
        tokens_li.append(''.join([i for i in token if not i.isdigit()]))
    tokens_str = ' '.join(tokens_li)
    tokens_str = re.sub(dil," ",tokens_str.lower())
    tokens_li = tokens_str.split()
    tokens_li = list(filter(None, tokens_li))
    ori_tokens_num = len(tokens_li)
    tokens_li = [i for i in tokens_li if i in glove_words]
    try:
        rate = len(tokens_li)/ori_tokens_num
    except ZeroDivisionError:
        rate = np.nan
    return tokens_li , rate#ori_tokens_num/len(tokens_li)
        
    

In [143]:
# train_df = pd.DataFrame(columns=['id','category','text','original_tokenize_num','preprocess_text','preprocss_tokenize_num'])
train_df = pd.DataFrame(columns=['id','category','text','original_tokenize_num',
                                 'text_keras','rate_keras','text_nltk','rate_nltk','text_apacy','rate_spacy'])
train_dir = './data/20news-bydate-v3/20news-bydate-train/'
cat_dir = next(os.walk(train_dir))[1]
count = 0
for cat in tqdm(cat_dir):
    in_dir = train_dir + cat + '/'
    news_list = next(os.walk(in_dir))[2]
    news_list = [os.path.join(in_dir, f) for f in news_list]
    news_list = list(filter(lambda f: f.endswith(".txt"), news_list))
    for news_path in news_list:
        id_num = news_path.split('/')[-1].split('.')[0]
        with open(news_path,'r',encoding='latin1') as f:
            news_text = f.read()
            news_text = re.sub('\n',' ',news_text)
            news_text_tok_ori = T.text_to_word_sequence(news_text,filters='',lower=False)
            news_text_ori = " ".join(news_text_tok_ori)
            news_text_tok_num_ori = len(news_text_tok_ori)
            news_text_tok_keras,rate_keras = text_preprcess(news_text,tokenizers='keras')
            news_text_tok_nltk,rate_nltk = text_preprcess(news_text,tokenizers='nltk')
            news_text_tok_spacy,rate_spacy = text_preprcess(news_text,tokenizers='spacy')
        train_df.loc[count] = [id_num,cat,news_text_ori,news_text_tok_num_ori,
                               news_text_tok_keras,rate_keras,news_text_tok_nltk,rate_nltk,news_text_tok_spacy,rate_spacy]
        count+=1
#         break
    
# train_df.loc[0] = [0,0,0]
print(train_df.rate_keras.mean() , train_df.rate_nltk.mean(),train_df.rate_spacy.mean()  )
train_df

100%|██████████| 20/20 [03:53<00:00, 11.71s/it]

0.9595051529116455 0.959558811893717 0.9605178211130393


id          category  \
0      19305  rec.sport.hockey   
1      19313  rec.sport.hockey   
2      19122  rec.sport.hockey   
3      19312  rec.sport.hockey   
4      19115  rec.sport.hockey   
...      ...               ...   
11309  20797       alt.atheism   
11310  20537       alt.atheism   
11311  20757       alt.atheism   
11312  20677       alt.atheism   
11313  20719       alt.atheism   

                                                    text  \
0      My votes (FWIW): Team MVP: Pat Verbeek. He fan...   
1      Kovalev is too talented a player to play for R...   
2      New Jersey 1 0 2--3 Pittsburgh 2 3 1--6 First ...   
3      First of all, the Penguins WILL win the cup ag...   
4                       You can't. But good luck trying.   
...                                                  ...   
11309  : When the object of their belief is said to b...   
11310  Koff! You mean that as long as I put you to sl...   
11311  : Mr Connor's assertion that "more complex" ==...   
11312  (excess stuff deleted...) I know of a similar ...   
11313  The "System" refered to a "moral system". You ...   

      original_tokenize_num  \
0                       453   
1                       160   
2                       919   
3                        66   
4                         6   
...                     ...   
11309                   269   
11310                    23   
11311                   135   
11312                   139   
11313                    40   

                                              text_keras  rate_keras  \
0      [my, votes, fwiw, team, mvp, pat, he, fans, go...    0.950207   
1      [kovalev, talented, player, play, roger, niels...    0.951220   
2      [new, jersey, pittsburgh, first, period, pitts...    0.867299   
3      [first, all, penguins, will, win, cup, again, ...    0.981481   
4                  [you, can't, but, good, luck, trying]    1.000000   
...                                                  ...         ...   
11309  [when, object, belief, said, perfect, make, be...    0.930233   
11310  [koff, you, mean, long, i, sleep, first, i, ki...    1.000000   
11311  [mr, assertion, more, complex, later, paleonto...    0.971014   
11312  [excess, stuff, deleted, i, know, similar, inc...    0.987805   
11313  [the, system, refered, moral, system, you, sho...    0.950000   

                                               text_nltk  rate_nltk  \
0      [my, votes, fwiw, team, mvp, pat, he, fans, go...   0.950207   
1      [kovalev, talented, player, play, roger, needs...   0.938272   
2      [new, jersey, pittsburgh, first, period, pitts...   0.867299   
3      [first, all, penguins, will, win, cup, again, ...   0.981481   
4                [you, ca, n't, but, good, luck, trying]   1.000000   
...                                                  ...        ...   
11309  [when, object, belief, said, perfect, make, be...   0.978102   
11310  [koff, you, mean, long, i, sleep, first, i, ki...   1.000000   
11311  [mr, connor, 's, assertion, more, complex, lat...   0.985915   
11312  [excess, stuff, deleted, i, know, similar, inc...   0.987952   
11313  [the, system, refered, moral, system, you, hav...   1.000000   

                                              text_apacy  rate_spacy  
0      [my, votes, fwiw, team, mvp, pat, he, fans, go...    0.950207  
1      [kovalev, talented, player, play, roger, needs...    0.938272  
2      [new, jersey, pittsburgh, first, period, pitts...    0.867299  
3      [first, all, penguins, will, win, cup, again, ...    0.981481  
4                [you, ca, n't, but, good, luck, trying]    1.000000  
...                                                  ...         ...  
11309  [when, object, belief, said, perfect, make, be...    0.978102  
11310  [koff, you, mean, long, i, sleep, first, i, ki...    1.000000  
11311  [mr, connor, 's, assertion, more, complex, lat...    0.985915  
11312  [excess, stuff, deleted, i, know, similar, inc...    0.987952  
11313  [

* keras tokenizer: 95.95% tokens in glove
* nltk tokenizer: 95.96% tokens in glove
* spacy tokenizer: 96.05% tokens in glove

In [144]:
train_df = pd.DataFrame(columns=['id','category','text','original_tokenize_num','preprocess_text','preprocss_tokenize_num'])
train_dir = './data/20news-bydate-v3/20news-bydate-train/'
cat_dir = next(os.walk(train_dir))[1]
count = 0
for cat in tqdm(cat_dir):
    in_dir = train_dir + cat + '/'
    news_list = next(os.walk(in_dir))[2]
    news_list = [os.path.join(in_dir, f) for f in news_list]
    news_list = list(filter(lambda f: f.endswith(".txt"), news_list))
    for news_path in news_list:
        id_num = news_path.split('/')[-1].split('.')[0]
        with open(news_path,'r',encoding='latin1') as f:
            news_text = f.read()
            news_text = re.sub('\n',' ',news_text)
            news_text_tok_ori = T.text_to_word_sequence(news_text,filters='',lower=False)
            news_text_ori = " ".join(news_text_tok_ori)
            news_text_tok_num_ori = len(news_text_tok_ori)
#             news_text_tok_pre,rate_keras = text_preprcess(news_text,tokenizers='keras')
#             news_text_tok_pre,rate_nltk = text_preprcess(news_text,tokenizers='nltk')
            news_text_tok_pre,rate_spacy = text_preprcess(news_text,tokenizers='spacy')
            news_text_pre = " ".join(news_text_tok_pre)
        train_df.loc[count] = [id_num,cat,news_text_ori,news_text_tok_num_ori,news_text_pre,len(news_text_tok_pre)]
        count+=1
        break
    
train_df

100%|██████████| 20/20 [00:00<00:00, 42.82it/s]


id                  category  \
0   19305          rec.sport.hockey   
1   21002             comp.graphics   
2   21586        talk.politics.guns   
3   23146        talk.politics.misc   
4   23229     talk.politics.mideast   
5   19586                 sci.space   
6   22675                 sci.crypt   
7   27525  comp.sys.ibm.pc.hardware   
8   26073        rec.sport.baseball   
9   28486           sci.electronics   
10  25314              misc.forsale   
11  17681           rec.motorcycles   
12  25854        talk.religion.misc   
13  23931                   sci.med   
14  26597     comp.sys.mac.hardware   
15  18684    soc.religion.christian   
16  20307            comp.windows.x   
17  24375   comp.os.ms-windows.misc   
18  28065                 rec.autos   
19  20556               alt.atheism   

                                                 text original_tokenize_num  \
0   My votes (FWIW): Team MVP: Pat Verbeek. He fan...                   453   
1   I have posted disp135.zip to alt.binaries.pict...                  1447   
2   Why don't you call the City and ask? Oak Park ...                   327   
3   -> >Now let me get this straight. After a nice...                   301   
4   NEWS YOU MAY HAVE MISSED, APR 19, 1993 Not bec...                   615   
5   Is this the one that had the {wrench|pliers} f...                    12   
6   Hmm, followup on my own posting... Well, who c...                   973   
7   : I'm looking to buy a 17" monitor soon, and i...                   111   
8   Methinks you recall wrong. Mitchell hit close ...                    84   
9   Fortunately, wire-wrapping is a better wiring ...                   125   
10  Forsale: SONY MHC-3600 HI-FI Bookshelf stereo ...                   120   
11  Paraphrase of initial post: \tCan I fight a sp...                    49   
12  Please, please don't make Barney to a modern m...                    33   
13  A relative of mine has recently been diagnosed...                    43   
14                                                                        0   
15  I apologize if this post isn't entirely approp...                    52   
16  I have heard of two packages for the PC that s...                    52   
17  What exactly does the windows bitmap format lo...                    43   
18  I'm wondering if anybody else out there is a c...                   166   
19  Just what do gay people do that straight peopl...                   128   

                                      preprocess_text preprocss_tokenize_num  
0   my votes fwiw team mvp pat he fans goal mouth ...                    229  
1   i posted you distribute program freely noncomm...                    912  
2   why city ask oak park illegal handgun ban well...                    180  
3   now let straight after nice long rant how peop...                    150  
4   news you may have missed apr not busy us media...                    368  
5                                  is inside recovery                      3  
6   hmm followup posting well cares first let try ...                    481  
7   i 'm looking buy monitor soon i ca n't decide ...                     56  
8   methinks recall wrong mitchell hit close atlan...                     39  
9   fortunately wirewrapping better wiring techniq...                     71  
10  forsale sony mhc hifi bookshelf stereo months ...                     80  
11  paraphrase initial post can i fight speeding t...                     35  
12  please make barney modern mythical figure i de...                     17  
13  a relative recently diagnosed stage papillary ...                     23  
14                                                                         0  
15  i apologize post entirely appropriate newsgrou...                     26  
16  i heard packages pc support xwin the linux fre...                     23  
17  what exactly windows bitmap format look like i...                     24  
18  i 'm wondering anybody

In [207]:
train_df.loc[1,'text']

'I have posted disp135.zip to alt.binaries.pictures.utilities ****** You may distribute this program freely for non-commercial use if no fee is gained. ****** There is no warranty. The author is not responsible for any damage caused by this program. Important changes since version 1,30: Fix bugs in file management system (file displaying). Improve file management system (more user-friendly). Fix bug in XPM version 3 reading. Fix bugs in TARGA reading/writng. Fix bug in GEM/IMG reading. Add support for PCX and GEM/IMG writing. Auto-skip macbinary header. (1) Introduction: This program can let you READ, WRITE and DISPLAY images with different formats. It also let you do some special effects(ROTATION, DITHERING ....) on image. Its main purpose is to let you convert image among different formts. Include simple file management system. Support \'slide show\'. There is NO LIMIT on image size. Currently this program supports 8, 15, 16, 24 bits display. If you want to use HiColor or TrueColor, 

In [145]:
def basic_statistics(all_length):
    '''
    input: length list of elements e.g.[1,1,1,3,5,9,4,2,1,3,54,78,5...]
    output1: mean、std、mode、min、q1、median(q2)、q3、max、iqr、outlier、far out
    output2: statistics graph、10%~90% form
    '''
    stat_dict = {}
    stat_dict['mean'] = np.mean(all_length)
    stat_dict['std'] = np.std(all_length)
    stat_dict['mode'] = np.argmax(np.bincount(all_length))
    stat_dict['min'] = np.min(all_length)
    stat_dict['q1'] = np.quantile(all_length,0.25)
    stat_dict['median'] = np.quantile(all_length,0.5)
    stat_dict['q3'] = np.quantile(all_length,0.75)
    stat_dict['max'] = np.max(all_length)
    stat_dict['iqr'] = stat_dict['q3'] - stat_dict['q1']
    stat_dict['outlier'] = stat_dict['q3'] + 1.5*stat_dict['iqr']
    stat_dict['far_out'] = stat_dict['q3'] + 3*stat_dict['iqr']
    for i in [10,20,30,40,50,60,70,80,90,100]:
        stat_dict[str(i)+'%'] = np.percentile(all_length,i)
    return pd.DataFrame.from_dict(stat_dict,orient='index',columns=['length'])

In [147]:
tokens_num_li = train_df.preprocss_tokenize_num.tolist()
basic_statistics(tokens_num_li)

length
mean     144.450000
std      214.881008
mode      23.000000
min        0.000000
q1        23.750000
median    63.500000
q3       157.500000
max      912.000000
iqr      133.750000
outlier  358.125000
far_out  558.750000
10%       15.600000
20%       23.000000
30%       25.400000
40%       37.400000
50%       63.500000
60%       77.600000
70%      112.200000
80%      189.800000
90%      379.300000
100%     912.000000

In [45]:
news_text_tok

['just',
 'what',
 'do',
 'gay',
 'people',
 'do',
 'that',
 'straight',
 'people',
 "don't?",
 'absolutely',
 'nothing.',
 "i'm",
 'a',
 'very',
 'straight(as',
 'an',
 'arrow),',
 '17-year',
 'old',
 'male',
 'that',
 'is',
 'involved',
 'in',
 'the',
 'bsa.',
 'i',
 "don't",
 'care',
 'what',
 'gay',
 'people',
 'do',
 'among',
 'each',
 'other,',
 'as',
 'long',
 'as',
 'they',
 "don't",
 'make',
 'passes',
 'at',
 'me',
 'or',
 'anything.',
 'at',
 'my',
 'summer',
 'camp',
 'where',
 'i',
 'work,',
 'my',
 'boss',
 'is',
 'gay.',
 'not',
 'in',
 'a',
 "'pansy'",
 'way',
 'of',
 'gay',
 '(i',
 'know',
 'a',
 'few),',
 'but',
 'just',
 "'one",
 'of',
 'the',
 "guys'.",
 'he',
 "doesn't",
 'push',
 'anything',
 'on',
 'me,',
 'and',
 'we',
 'give',
 'him',
 'the',
 'same',
 'respect',
 'back,',
 'due',
 'to',
 'his',
 'position.',
 'if',
 'anything,',
 'the',
 'bsa',
 'has',
 'taught',
 'me,',
 'i',
 "don't",
 'know,',
 'tolerance',
 'or',
 'something.',
 'before',
 'i',
 'met',
 't

In [30]:
news_list = next(os.walk(in_dir))[2]
news_list = [os.path.join(in_dir, f) for f in news_list]
news_list

['./data/20news-bydate-v3/20news-bydate-train/alt.atheism/20556.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20748.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20810.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20574.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20692.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20705.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20519.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20565.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20835.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20649.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20684.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20872.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20739.txt',
 './data/20news-bydate-v3/20news-bydate-train/alt.atheism/20547.txt',
 './data/20news-byda

# TEST

In [203]:
import torch

In [204]:
roberta = torch.hub.load('pytorch/fairseq', 'roberta.large')
roberta.eval() 

Using cache found in /home/leoqaz12/.cache/torch/hub/pytorch_fairseq_master


loading archive file http://dl.fbaipublicfiles.com/fairseq/models/roberta.large.tar.gz from cache at /home/leoqaz12/.cache/torch/pytorch_fairseq/83e3a689e28e5e4696ecb0bbb05a77355444a5c8a3437e0f736d8a564e80035e.c687083d14776c1979f3f71654febb42f2bb3d9a94ff7ebdfe1ac6748dba89d2
| dictionary: 50264 types


RobertaHubInterface(
  (model): RobertaModel(
    (decoder): RobertaEncoder(
      (sentence_encoder): TransformerSentenceEncoder(
        (embed_tokens): Embedding(50265, 1024, padding_idx=1)
        (embed_positions): LearnedPositionalEmbedding(514, 1024, padding_idx=1)
        (layers): ModuleList(
          (0): TransformerSentenceEncoderLayer(
            (self_attn): MultiheadAttention(
              (out_proj): Linear(in_features=1024, out_features=1024, bias=True)
            )
            (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
            (fc1): Linear(in_features=1024, out_features=4096, bias=True)
            (fc2): Linear(in_features=4096, out_features=1024, bias=True)
            (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
          )
          (1): TransformerSentenceEncoderLayer(
            (self_attn): MultiheadAttention(
              (out_proj): Linear(in_features=1024, out_features=1024, bias=Tru

In [206]:
tokens_ori

"now let straight after nice long rant how people need personal responsibility economic social lives sudden 's radicals such me i guess responsible poor people 's lifestyles tell think poor people dumb think themselves there reasons disintegration family and support systems general nation 's poor somehow i think murphy janis the sane person 's list you want generation 's vaunted cultural revolution lasting change worse try socalled relevant values education hey like good idea time how know needed real education mean took granted the 's generation spoiled irresponsible the depression create mothers fathers determined kids want going overboard creating nation brats consider contrast famous events july apollo woodstock which group large numbers people feed reverted cultural level primitives defecation public etc and group assembled took care itself dispersed damage deaths large numbers drug problems was n't woodstock called biggest parking lot history they rejected society went nature par

In [205]:
tokens_ori = train_df.loc[3,'preprocess_text']#'how\'s are you'#
tokens = roberta.encode(tokens_ori)
# assert tokens.tolist() == [0, 31414, 232, 328, 2]
# assert roberta.decode(tokens) == 'Hello world!'
print(len(tokens) , tokens)
roberta.decode(tokens)
doc = roberta.extract_features_aligned_to_words(tokens_ori)
print(len(doc))
for tok in doc:
    print('{:10}{} (...)'.format(str(tok), tok.vector[:5]))

173 tensor([    0,  8310,   905,  1359,    71,  2579,   251, 25693,   141,    82,
          240,  1081,  2640,   776,   592,  1074,  7207,   128,    29, 35842,
          215,   162,   939,  4443,  2149,  2129,    82,   128,    29, 28182,
         1137,   206,  2129,    82, 16881,   206,  1235,    89,  2188, 32654,
         8475,   284,     8,   323,  1743,   937,  1226,   128,    29,  2129,
         7421,   939,   206, 22802, 16628, 10408,   354,     5, 37091,   621,
          128,    29,   889,    47,   236,  2706,   128,    29,   748, 19264,
         4106,  7977,  9735,   464,  3007,   860, 17380,  9315,  4249,  3266,
         1265, 17232,   101,   205,  1114,    86,   141,   216,   956,   588,
         1265,  1266,   362,  4159,     5,   128,    29,  2706, 29136, 21573,
            5,  6943,  1045,  8826, 17850,  3030,  1159,   236,   164, 35912,
         2351,  1226,  5378,  2923,  1701,  5709,  3395,  1061,  1236, 21347,
         6256, 38520,  5627,  9607,    61,   333,   739,  15

AssertionError: 

In [189]:
tokens = 'I said, "hello RoBERTa."'
tokens = sp_tokenizer(tokens)
token_li = []
for token in tokens:
    token_li.append(token.text)
len(token_li) , token_li

(8, ['I', 'said', ',', '"', 'hello', 'RoBERTa', '.', '"'])

In [167]:
last_layer_features = roberta.extract_features(tokens)
print(last_layer_features.size())# == torch.Size([1, 5, 1024])

all_layers = roberta.extract_features(tokens, return_all_hiddens=True)
assert len(all_layers) == 25
assert torch.all(all_layers[-1] == last_layer_features)

ValueError: tokens exceeds maximum length: 592 > 512

In [163]:
last_layer_features.detach().numpy().shape

(1, 19, 1024)

In [168]:
from pytorch_transformers import RobertaTokenizer

In [173]:
tokenizer = RobertaTokenizer.from_pretrained('roberta-large-mnli')

100%|██████████| 456318/456318 [00:00<00:00, 545019.70B/s]


In [178]:
len(tokenizer.encode(train_df.loc[0,'text']))

W0925 15:40:29.729804 139833327294272 tokenization_utils.py:665] Token indices sequence length is longer than the specified maximum sequence length for this model (590 > 512). Running this sequence through the model will result in indexing errors


590